In [1]:
import pandas as pd

last_month_billing = pd.read_csv('last_month_billing.csv')
current_customers = pd.read_csv('current_customers.csv')
all_time_billing = pd.read_csv('all_time_billing.csv')

In [2]:
last_month_billing.head()

Company ID  Total Billing Across Portfolio Last Month  Marvin Gardens  \
0   460881659                                   0.020384        0.021789   
1   112755124                                   0.012646        0.000000   
2    18059723                                   0.002704        0.000000   
3   350690094                                   0.004674        0.000000   
4    20054213                                   0.016864        0.000000   

   Park Place  Boardwalk   Ventnor  St. James Place  Tennessee  Kentucky  \
0    0.000000   0.000000  0.000000         0.000000   0.000000       0.0   
1    0.048441   0.000000  0.000000         0.000000   0.055568       0.0   
2    0.000000   0.020537  0.000000         0.000000   0.000000       0.0   
3    0.000000   0.000000  0.023119         0.000000   0.000000       0.0   
4    0.000000   0.000000  0.000000         0.089201   0.000000       0.0   

   Pennslyvania  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

In [3]:
current_customers.head()

Company ID  Founded Year  Revenue (in 000s USD) Revenue Range (in USD)  \
0     1475361        1994.0              538046000           Over $5 bil.   
1    16400573        1998.0              256735000           Over $5 bil.   
2      246280        1989.0               63144419           Over $5 bil.   
3    43076614        1999.0              364115000           Over $5 bil.   
4   126828694        1983.0              136835000           Over $5 bil.   

   Est. Marketing Department Budget (in 000s USD)  \
0                                      69407934.0   
1                                      33118815.0   
2                                       4862120.0   
3                                      18205750.0   
4                                      11083635.0   

   Est. Finance Department Budget (in 000s USD)  \
0                                     5380460.0   
1                                     2567350.0   
2                                      631444.0   
3                                     3641150.0   
4                                     2052525.0   

   Est. IT Department Budget (in 000s USD)  \
0                               15065288.0   
1                                7188580.0   
2                                2967787.0   
3                               10559335.0   
4                                5062895.0   

   Est. HR Department Budget (in 000s USD)  Employees       Employee Range  \
0                                4041427.0    1541000  Employees.10000plus   
1                                 228090.0     160000  Employees.10000plus   
2                                1663794.0     738000  Employees.10000plus   
3                                 139776.0      62000  Employees.10000plus   
4                                 253627.0     112500  Employees.10000plus   

   ...     Recent Funding Round  Recent Funding Date  Recent Investors  \
0  ...  Stock Issuance/Offering            14-May-97               NaN   
1  ...  Stock Issuance/Offering            19-Aug-04       Undisclosed   
2  ...  Stock Issuance/Offering            18-Jul-01               NaN   
3  ...                      NaN                  NaN               NaN   
4  ...                      NaN                  NaN               NaN   

                                       All Investors   Company City  \
0                                                NaN        Seattle   
1  Andy Bechtolsheim;Undisclosed;Kleiner Perkins ...  Mountain View   
2                                                NaN         Dublin   
3                                                NaN         Spring   
4                                                NaN  New York City   

   Company State  Company Zip Code Company Country Number of Locations  \
0     Washington             98109   United States                37.0   
1     California             94043   United States                91.0   
2       Leinster          D02 P820         Ireland                49.0   
3          Texas             77389   United States                28.0   
4       New York             10036   United States                55.0   

  Company Is Acquired  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  

[5 rows x 40 columns]

In [4]:
all_time_billing.head()

Company ID  Total Billing Across Portfolio Last Month  Marvin Gardens  \
0   387494486                                   0.000350        0.000000   
1    97912844                                   0.000784        0.000000   
2    46925340                                   0.000770        0.000896   
3   368572921                                   0.362776        0.079741   
4   126828694                                   0.445493        0.203161   

   Park Place  Boardwalk   Ventnor  St. James Place  Tennessee  Kentucky  \
0    0.000000   0.000000  0.000000         0.000000   0.006811  0.000000   
1    0.000000   0.000000  0.000000         0.000000   0.000000  0.017075   
2    0.000000   0.000000  0.000000        -0.000318   0.000000 -0.001258   
3    0.027441   0.226838  0.867738         0.949844   0.604594  0.192237   
4    1.000000   0.287801  0.316198         0.348450   0.381109  0.126162   

   Pennslyvania  Unnamed: 10  Unnamed: 11  Unnamed: 12  Unnamed: 13  \
0           0.0          NaN          NaN          NaN          NaN   
1           0.0          NaN          NaN          NaN          NaN   
2           0.0          NaN          NaN          NaN          NaN   
3           0.0          NaN          NaN          NaN          NaN   
4           0.0          NaN          NaN          NaN          NaN   

   Unnamed: 14  Unnamed: 15  Unnamed: 16  Unnamed: 17  Unnamed: 18  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 19  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN

In [5]:
all_time_combined = pd.merge(all_time_billing, current_customers, on='Company ID', how='inner')
last_month_combined = pd.merge(last_month_billing, current_customers, on='Company ID', how='inner')

In [6]:
all_time_combined.head()

Company ID  Total Billing Across Portfolio Last Month  Marvin Gardens  \
0   387494486                                   0.000350        0.000000   
1    97912844                                   0.000784        0.000000   
2    46925340                                   0.000770        0.000896   
3   368572921                                   0.362776        0.079741   
4   126828694                                   0.445493        0.203161   

   Park Place  Boardwalk   Ventnor  St. James Place  Tennessee  Kentucky  \
0    0.000000   0.000000  0.000000         0.000000   0.006811  0.000000   
1    0.000000   0.000000  0.000000         0.000000   0.000000  0.017075   
2    0.000000   0.000000  0.000000        -0.000318   0.000000 -0.001258   
3    0.027441   0.226838  0.867738         0.949844   0.604594  0.192237   
4    1.000000   0.287801  0.316198         0.348450   0.381109  0.126162   

   Pennslyvania  ...  Recent Funding Round  Recent Funding Date  \
0           0.0  ...                   NaN                  NaN   
1           0.0  ...                   NaN                  NaN   
2           0.0  ...                  Debt                43935   
3           0.0  ...                   NaN                  NaN   
4           0.0  ...                   NaN                  NaN   

   Recent Investors  All Investors   Company City  Company State  \
0               NaN            NaN        De Soto         Kansas   
1               NaN            NaN     Warrington   Pennsylvania   
2               PPP            PPP     Wilmington       Delaware   
3               NaN            NaN  New York City       New York   
4               NaN            NaN  New York City       New York   

   Company Zip Code  Company Country  Number of Locations  Company Is Acquired  
0             66018    United States                  1.0                  NaN  
1             18976    United States                  7.0                  Yes  
2             19801    United States                  NaN                  NaN  
3             10019    United States                 10.0                  NaN  
4             10036    United States                 55.0                  NaN  

[5 rows x 59 columns]

In [7]:
last_month_combined.head()

Company ID  Total Billing Across Portfolio Last Month  Marvin Gardens  \
0   460881659                                   0.020384        0.021789   
1   460881659                                   0.020384        0.021789   
2   460881659                                   0.020384        0.021789   
3   460881659                                   0.020384        0.021789   
4   460881659                                   0.020384        0.021789   

   Park Place  Boardwalk  Ventnor  St. James Place  Tennessee  Kentucky  \
0         0.0        0.0      0.0              0.0        0.0       0.0   
1         0.0        0.0      0.0              0.0        0.0       0.0   
2         0.0        0.0      0.0              0.0        0.0       0.0   
3         0.0        0.0      0.0              0.0        0.0       0.0   
4         0.0        0.0      0.0              0.0        0.0       0.0   

   Pennslyvania  ...   Recent Funding Round  Recent Funding Date  \
0           0.0  ...  Other Financing Round                44438   
1           0.0  ...  Other Financing Round                44438   
2           0.0  ...  Other Financing Round                44438   
3           0.0  ...  Other Financing Round                44438   
4           0.0  ...  Other Financing Round                44438   

                                    Recent Investors  \
0  Co-Win Ventures;Oriza Seed;InnoAngel;Tsingyuan...   
1  Co-Win Ventures;Oriza Seed;InnoAngel;Tsingyuan...   
2  Co-Win Ventures;Oriza Seed;InnoAngel;Tsingyuan...   
3  Co-Win Ventures;Oriza Seed;InnoAngel;Tsingyuan...   
4  Co-Win Ventures;Oriza Seed;InnoAngel;Tsingyuan...   

                                       All Investors   Company City  \
0  Co-Win Ventures;Oriza Seed;InnoAngel;Tsingyuan...  State College   
1  Co-Win Ventures;Oriza Seed;InnoAngel;Tsingyuan...  State College   
2  Co-Win Ventures;Oriza Seed;InnoAngel;Tsingyuan...  State College   
3  Co-Win Ventures;Oriza Seed;InnoAngel;Tsingyuan...  State College   
4  Co-Win Ventures;Oriza Seed;InnoAngel;Tsingyuan...  State College   

   Company State  Company Zip Code  Company Country Number of Locations  \
0   Pennsylvania             16803    United States                 2.0   
1   Pennsylvania             16803    United States                 2.0   
2   Pennsylvania             16803    United States                 2.0   
3   Pennsylvania             16803    United States                 2.0   
4   Pennsylvania             16803    United States                 2.0   

   Company Is Acquired  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  

[5 rows x 49 columns]

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

In [9]:
target = 'Total Billing Across Portfolio Last Month'
features = last_month_combined.select_dtypes(include=[np.number]).columns.tolist()
features.remove(target)

numeric_columns = last_month_combined.select_dtypes(include=[np.number]).columns
last_month_combined[numeric_columns] = last_month_combined[numeric_columns].fillna(last_month_combined[numeric_columns].mean())


In [10]:
X_train, X_test, y_train, y_test = train_test_split(
    last_month_combined[features],
    last_month_combined[target],
    test_size=0.2,
    random_state=42
)


In [11]:
models = {
    "Linear Regression": LinearRegression(),
    "Random Forest": RandomForestRegressor(random_state=42)
}

results = {}

for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    results[name] = {"MSE": mse, "R^2": r2}

results



{'Linear Regression': {'MSE': 9.826817926943617e-17,
  'R^2': 0.9999999999999966},
 'Random Forest': {'MSE': 1.080921328027574e-08, 'R^2': 0.999999621893678}}

## Decision Tree Regressor

In [12]:
from sklearn.tree import DecisionTreeRegressor
TreeReg = DecisionTreeRegressor()
TreeReg.fit(X_train, y_train)

y_pred_Train = TreeReg.predict(X_train) #Predictions on training set
y_pred_Test = TreeReg.predict(X_test)   #Predictions on testing set

## Train Metrics

In [13]:
from sklearn import metrics
print('R squared:', metrics.r2_score(y_train, y_pred_Train))

R squared: 1.0


In [14]:
print('Mean Absolute Error:', metrics.mean_absolute_error(y_train, y_pred_Train))  

Mean Absolute Error: 4.565791012524656e-17


In [15]:
print('Mean Squared Error:', metrics.mean_squared_error(y_train, y_pred_Train))  

Mean Squared Error: 1.4400698028499018e-32


In [16]:
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_train, y_pred_Train)))

Root Mean Squared Error: 1.2000290841683386e-16


## Test Metrics

In [17]:
print('R squared:', metrics.r2_score(y_test, y_pred_Test))

R squared: 0.9999999795953219


In [18]:
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred_Test))  

Mean Absolute Error: 1.3193260320587557e-06


In [19]:
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred_Test))  

Mean Squared Error: 5.833240655433867e-10


In [20]:
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred_Test)))

Root Mean Squared Error: 2.4152102714740734e-05


# KNN (K-Nearest Neighbors)

In [26]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error, r2_score

knn_model = KNeighborsRegressor()
knn_model.fit(X_train, y_train)
y_pred_knn = knn_model.predict(X_test)
mse_knn = mean_squared_error(y_test, y_pred_knn)
r2_knn = r2_score(y_test, y_pred_knn)

print("KNN Model:")
print(f"MSE: {mse_knn}")
print(f"R^2: {r2_knn}")


KNN Model:
MSE: 1.5668529785249866e-05
R^2: 0.9994519147679419


# Random Forest:

In [27]:
from sklearn.ensemble import RandomForestRegressor

rf_model = RandomForestRegressor(random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)
mse_rf = mean_squared_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)

print("Random Forest Model:")
print(f"MSE: {mse_rf}")
print(f"R^2: {r2_rf}")


Random Forest Model:
MSE: 1.080921328027574e-08
R^2: 0.999999621893678


# Extra Trees:

In [28]:
from sklearn.ensemble import ExtraTreesRegressor

et_model = ExtraTreesRegressor(random_state=42)
et_model.fit(X_train, y_train)
y_pred_et = et_model.predict(X_test)
mse_et = mean_squared_error(y_test, y_pred_et)
r2_et = r2_score(y_test, y_pred_et)

print("Extra Trees Model:")
print(f"MSE: {mse_et}")
print(f"R^2: {r2_et}")


Extra Trees Model:
MSE: 3.2793039873210603e-10
R^2: 0.9999999885289934


# Gradient Boosting Regressor (GBR):

In [29]:
from sklearn.ensemble import GradientBoostingRegressor

gbr_model = GradientBoostingRegressor(random_state=42)
gbr_model.fit(X_train, y_train)
y_pred_gbr = gbr_model.predict(X_test)
mse_gbr = mean_squared_error(y_test, y_pred_gbr)
r2_gbr = r2_score(y_test, y_pred_gbr)

print("Gradient Boosting Regressor Model:")
print(f"MSE: {mse_gbr}")
print(f"R^2: {r2_gbr}")


Gradient Boosting Regressor Model:
MSE: 6.742858744848177e-06
R^2: 0.9997641347752114


# XGBoost Regressor (XGBR):

In [30]:
from xgboost import XGBRegressor

xgbr_model = XGBRegressor(random_state=42)
xgbr_model.fit(X_train, y_train)
y_pred_xgbr = xgbr_model.predict(X_test)
mse_xgbr = mean_squared_error(y_test, y_pred_xgbr)
r2_xgbr = r2_score(y_test, y_pred_xgbr)

print("XGBoost Regressor Model:")
print(f"MSE: {mse_xgbr}")
print(f"R^2: {r2_xgbr}")


XGBoost Regressor Model:
MSE: 9.271673290938244e-08
R^2: 0.9999967567683268


# CatBoost

In [31]:
from catboost import CatBoostRegressor

catboost_model = CatBoostRegressor(random_state=42, verbose=0)
catboost_model.fit(X_train, y_train)
y_pred_catboost = catboost_model.predict(X_test)
mse_catboost = mean_squared_error(y_test, y_pred_catboost)
r2_catboost = r2_score(y_test, y_pred_catboost)

print("CatBoost Model:")
print(f"MSE: {mse_catboost}")
print(f"R^2: {r2_catboost}")


CatBoost Model:
MSE: 2.227094659726628e-08
R^2: 0.9999992209622025


# LightGBM:

In [33]:
from lightgbm import LGBMRegressor

lgbm_model = LGBMRegressor(random_state=42)
lgbm_model.fit(X_train, y_train)
y_pred_lgbm = lgbm_model.predict(X_test)
mse_lgbm = mean_squared_error(y_test, y_pred_lgbm)
r2_lgbm = r2_score(y_test, y_pred_lgbm)

print("LightGBM Model:")
print(f"MSE: {mse_lgbm}")
print(f"R^2: {r2_lgbm}")


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.001766 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 3281
[LightGBM] [Info] Number of data points in the train set: 4747, number of used features: 25
[LightGBM] [Info] Start training from score 0.089100
LightGBM Model:
MSE: 4.697715027192386e-07
R^2: 0.9999835673910309


# Evaluation of all models:

| Model                        | MSE                 | R^2                  |
|------------------------------|---------------------|----------------------|
| Decision Tree Regressor      | Train: 1.440e-32    Test: 5.833e-10       | Train: 1.0       Test: 0.999999979 |
| KNN                          | 1.567e-05           | 0.9995               |
| Random Forest                | 1.081e-08           | 0.99999962           |
| Extra Trees                  | 3.279e-10           | 0.99999999           |
| Gradient Boosting Regressor  | 6.743e-06           | 0.9998               |
| XGBoost Regressor            | 9.272e-08           | 0.999997             |
| CatBoost                     | 2.227e-08           | 0.9999992            |
| LightGBM                     | 4.698e-07           | 0.999984             |


# Analysis:
## Decision Tree Regressor:
The model achieves perfect performance on the training set but slightly lower performance on the test set, indicating potential overfitting.
## KNN (K-Nearest Neighbors):
KNN performs well with a low MSE and high R^2 score, suggesting it's a good fit for the data.
## Random Forest:
Random Forest performs exceptionally well, with very low MSE and high R^2 score on both the training and test sets.
## Extra Trees:
Extra Trees also performs very well, with extremely low MSE and high R^2 score, indicating it's capturing the underlying patterns effectively.
## Gradient Boosting Regressor (GBR):
GBR shows good performance but slightly higher MSE compared to Random Forest and Extra Trees.
## XGBoost Regressor (XGBR):
XGBR demonstrates strong performance with very low MSE and high R^2 score, similar to Random Forest.
## CatBoost:
CatBoost performs well with low MSE and high R^2 score, indicating it's effectively capturing the relationship between features and target.
## LightGBM:
LightGBM performs well but with slightly higher MSE compared to other models, suggesting it may not be the best fit for this dataset.

#### Overall, Random Forest, Extra Trees, XGBoost, and CatBoost are the top-performing models based on MSE and R^2 scores. These models are well-suited for this regression task and can provide accurate predictions for total billing across the portfolio last month.